diff --git a/src/pkcs11/mechanism.c b/src/pkcs11/mechanism.c
index 6b879fa52..aaa0017db 100644
--- a/src/pkcs11/mechanism.c
+++ b/src/pkcs11/mechanism.c
@@ -203,6 +203,9 @@ sc_pkcs11_find_mechanism(struct sc_pkcs11_card *p11card, CK_MECHANISM_TYPE mech,
 	sc_pkcs11_mechanism_type_t *mt;
 	unsigned int n;
 
+
+	if (p11card == NULL)
+		return NULL;
 	for (n = 0; n < p11card->nmechanisms; n++) {
 		mt = p11card->mechanisms[n];
 		if (mt && mt->mech == mech && ((mt->mech_info.flags & flags) == flags))
